### ISSUE TITLE
Opacity with line 2d and anti-aliasing on causes sprite debug boxes to appear

### ISSUE THREAD

**Issue by EricEzaM** [godotengine/godot#35878]
Not sure if this is as it supposed to be or not... But might as well try submit a bug to see if it is one.

**Godot version:**
3.2 stable
**OS/device including version:**
Windows 10 1903
**Issue description:**
Line 2d shows debug boxes when anti-aliasing is enabled and material is translucent (i.e. 0 < alpha < 1). Example below
![Godot-Line-Bug](https://user-images.githubusercontent.com/41730826/73651260-cd9cc780-46cf-11ea-90a2-9001fa793977.png)

**Steps to reproduce:**
New project
Add Line2d
Add some points
Set default color alpha to be <1, e.g. 0.5
Make width larger (to better see effect), e.g. 30 - 50
Enable anti-aliasing under 'border'

**Comment by YeldhamDev** [comment]
I think it's the anti-aliasing effect overlapping over the polygons rather than debug boxes.

**Comment by Calinou** [comment]
As @YeldhamDev said, I'm not sure if anything can be done about it.

**Comment by Zylann** [comment]
How is it possible that AA makes *triangles* of the line visible? They are even supposed to be shared vertices, there is no way they would overlap like that and show as if they were wireframe Oo

**Comment by SleepProgger** [comment]
Assuming the godot 3.2_stable release is the same as the current 3.2 branch:

Line2D antialiasing ends up in here: https://github.com/godotengine/godot/blob/669cd46495389916ebcb50bc00537b0402a0960c/drivers/gles3/rasterizer_canvas_gles3.cpp#L905 rendered as TYPE_POLYGON

`canvas_item_add_polyline` ends up being rendered as TYPE_POLYLINE ( https://github.com/godotengine/godot/blob/669cd46495389916ebcb50bc00537b0402a0960c/drivers/gles3/rasterizer_canvas_gles3.cpp#L690 ) which doesn't have this problem.

The reason seem to be that TYPE_POLYGON tries to smooth by drawing a line with smooth enabled between all indices while TYPE_POLYLINE draws the outline from the line attribute which is filled properly in `VisualServerCanvas::canvas_item_add_polyline`. As far as i see LineBuilder could be changed so it does the same and Line2D could then also draw via TYPE_POLYLINE in the end.
I may have missed something tho, its late for me.

Both implementation seem to have some overdraw regardless, which prevents them from being really useful with opaque:
![Untitled](https://user-images.githubusercontent.com/1274039/74081758-d1737580-4a52-11ea-94f6-a412d8df4e35.png)

**Comment by Zylann** [comment]
> _draw_generic_indices(GL_LINE_STRIP

I guess this is again an attempt at faking AA somehow? If it tries to draw a line on the edges of the polygon, it is doing a mistake by also drawing the inside triangles. It looks like a bug to me, for ANY kind of polygon, not just Line2D.
`Line2D` cannot use `add_polyline` instead because you'd basically throw away all features it currently has. Or maybe if AA is enabled, it could compute outlines in `LineBuilder` in addition to its polygons.
These kinds of problems remind me of https://github.com/godotengine/godot/issues/35279, I wish we simply had proper AA rather than workarounds like this...

The second example in your screenshot shows a different algorithm using round joints, so yeah you get different results. It will look more like `draw_polyline` if you turn off some settings.
Completely avoiding overlap in the geometry is hard in both cases, normally you'd get less if you had less points anyways. One way to get no overlap is to pre-draw the line opaque in a viewport and draw the result with transparency, but in current Godot it's quite overkill.

**Comment by SleepProgger** [comment]
>  Or maybe if AA is enabled, it could compute outlines in LineBuilder in addition to its polygons (which btw could allow to generate collision segments?).

Yes, that was the idea. Similar to draw_line, although i concur a better antialiasing strategy would be  the better approach.
I am currently writing a LineRenderer using a Mesh basically following https://blog.mapbox.com/drawing-antialiased-lines-with-opengl-8766f34192dc?gi=ea7cb7d50409
This uses the shader to blend the alpha on the line edges for antialiasing.
This would probably be not a good approach for godot tho as it relies on the shader (ie: AA would break if the user uses their own shader).
An alternative with the same effect would be to draw **6** triangles per line segment.

TBH. as is: i'd say line rendering with translucency is broken and this should at least be noted in the docs.

/kinda offtopic, sorry:
The overlapping problem is the same for all join types and i need custom color per vertex, so i guess i have to fiddle a bit more :)
![Untitled](https://user-images.githubusercontent.com/1274039/74090055-aff0a900-4aa7-11ea-9f21-8f36560d21fd.png)

**Comment by avencherus** [comment]
> pre-draw the geometry opaque in a viewport and draw the result with transparency

@Zylann I think I understand the first part is to use draw functions on a canvas, but what do you mean by draw the result with transparency?

**Comment by Calinou** [comment]
@avencherus It means you should render an opaque line to a canvas, and display the resulting ViewportTexture with a modulate value that makes it translucent.

**Comment by akien-mga** [comment]
Still seems reproducible in the current `3.x` branch. MRP: [TestLine2DAA.zip](https://github.com/godotengine/godot/files/8356056/TestLine2DAA.zip)

In `master` anti-aliasing doesn't seem to work so the bug is not reproducible, but might be if/when AA gets fixed.

**Comment by Calinou** [comment]
As a workaround, use https://github.com/godot-extended-libraries/godot-antialiased-line2d which supports translucent antialiased 2D lines.

**Comment by erayzesen** [comment]
Hi all. I encountered this issue in version 3.6, and the solution to the problem is actually quite simple. Godot applies antialiasing to the edges to smooth polygons. What causes this issue is the overlapping lines within the polygon. Managing these collective polygon drawings together and preventing the drawing of lines if 2 lines overlap will give the most accurate result. In my own application, I first draw the polygon with antialiasing disabled, and then I draw 1-pixel antialiased lines on the edges using draw_line. I didn't need an algorithm like checking for overlapping lines because it's a more specific task. I made exceptions for which lines should not be drawn, and my problem was resolved.

In other words, while the problem may seem complex, it arises from adjacent polygons that overlap at their edges. To address this, perhaps a new method that draws polygons collectively and manages these issues could be added. It could be something like draw_polygons(). Of course, you can also implement this method yourself via GDScript or cpp.
